[8.x] Adds class handling for Blade echo statements

lukeraymonddowning opened the pull request:
## What is it?

When working in Blade, I'll often want to always output a class in the same manner. I'll use Brick/Money as the example here, but this can apply to pretty much any class you can think of. 

Brick/Money is a final class, so I can't add a `Stringable` interface to it. Whenever I output `{{ $moneyObject }}`, I would want it to format as `£100.00`, for example. The same could be said for a `Carbon` instance. 

This PR adds a new `Blade::handle()` method that can be placed in the boot method of a Service Provider and allows the user to add intercepting closures for any class. The returned value will be outputted in Blade.

Think of this as the Exception Handler, but for Blade.

## Usage

```php
// AppServiceProvider
Blade::handle(Money::class, fn($object) => $object->formatTo('en_GB'));
Blade::handle(Carbon::class, fn($object) => $object->format('d/m/Y')));
```

```blade
<dl>
    <dt>Total</dt>
    <dd>{{ $product->total }}</dd> <!-- This is a money object, but will be outputted as an en_GB formatted string -->
    <dt>Created at</dt>
    <dd>{{ $product->created_at }}</dd> <!-- This is a Carbon object, but will be outputted as English date format -->
</dl>
```

## How does it work?

This works by wrapping a small piece of logic around php output at the parser level, which means that the object is evaluated just in time and can thus be treated as a standard object throughout the entire application; it is only transformed by the PHP script at the time of output.

Feel free to throw any questions back at me :-)

Thanks for all the hard work guys!

lukeraymonddowning committed:
Adds start of blade class handlers.

lukeraymonddowning committed:
Adds basis for blade class handlers.

lukeraymonddowning committed:
Adds facade method.

lukeraymonddowning committed:
Adds a bugfix for non-object values.

lukeraymonddowning committed:
Fixes tests and improves speed.

lukeraymonddowning committed:
Removes static property.

lukeraymonddowning commented:
This is a fixed version of #37477 which had broken unit tests. It also is faster than previously because it will completely skip the JIT check if no handlers have been specified.

lukeraymonddowning committed:
Fixed styling

lukeraymonddowning committed:
Refactors test

lukeraymonddowning committed:
Adds echo handler to escaped echos.

lukeraymonddowning committed:
Adds support for escaped echos.

lukeraymonddowning committed:
Small refactor

taylorotwell commented:
Does putting line breaks in the compiled echo syntax cause any unwanted side effects in regards to HTML spacing?

lukeraymonddowning commented:
Will check this tomorrow and add unit tests to prove

lukeraymonddowning committed:
Removes new lines from echo handling.

lukeraymonddowning committed:
Style fix

lukeraymonddowning commented:
@taylorotwell from testing in a real application, the whitespaces made no difference, but it doesn't really make much sense to keep the whitespaces there anyway so I've removed them. Also added a test in similar vain to those in the `BladeEchoTest` file that check for the preservation of newlines after the output.

lukeraymonddowning committed:
Bugfix

lukeraymonddowning committed:
Test fixes

lukeraymonddowning committed:
Simplifies logic.

lukeraymonddowning committed:
Fixes docblock

lukeraymonddowning committed:
Adds a test that actually executes the logic to test that it works correctly.

lukeraymonddowning committed:
Style fixes

driesvints commented on src/Illuminate/View/Compilers/BladeCompiler.php:
```diff
@@ -752,4 +759,14 @@ public function withoutComponentTags()
+     *
+     * @return void
+     */
+    public function handle(string $className, callable $handler)
```
Wondering if this should be `addEchoHandler`. `handle` is so generic and might be confused with a job's `handle` method. The former is a registration of a resolver while the later is an actual resolver.

lukeraymonddowning replied:
Agreed. I've renamed.

lukeraymonddowning committed:
Renames `handle` method

taylorotwell committed:
formatting

taylorotwell commented:
Renamed method to `stringable` and added `ReflectsClosures` trait to Blade compiler so that it is now possible to omit the class name as the first argument and just provide a type-hinted callback:

```php
Blade::stringable(fn (Money $money) => $money->formatTo('en_GB'));
```

lukeraymonddowning commented:
Love it @taylorotwell. Stringable fits in perfectly with the interface, so makes complete sense and the ability to type hint instead is pure fire 😎
